arch/M68K/M68KDisassembler.c

```diff
@@ -188,7 +188,7 @@ static uint64_t m68k_read_disassembler_64(const m68k_info *info, const uint64_t
 static unsigned int m68k_read_safe_16(const m68k_info *info, const uint64_t address)
 {
 	const uint64_t addr = (address - info->baseAddress) & info->address_mask;
-	if (info->code_len < 2) {
+	if (info->code_len < addr + 2) {
```

Comment:
So I'm a bit unsure this change (and the other read functions) why is addr included in the check? Address is a virtual address that doesn't to me have any affect if we should read the data or not but checking that we have some data to read is needed. Notice I changed this in https://github.com/aquynh/capstone/pull/1130 to just resolve some issue(s) that this is trying to fix. Did you fuzz with the latest code?

Reply:
I made this pull request from commit c6dbbd06f8283542dfbb1c88951bb8d2325b0605
so after the #1130 fix

Maybe the fix is wrong (I do not know about M68K) but the code is wrong as well for now, as it can produce a out of bounds read.
`m68k_read_safe_16`calls `m68k_read_disassembler_16`which uses `info->code[addr + 0]`

IMHO, either
- we have to check `info->code_len`versus `addr+2`
- m68k_read_disassembler_16 should use `info->code[0]` rather than `info->code[addr + 0]`

Reply:
@aquynh do you want me to separate the different commits in different pull requests ?

I pushed commits on this branch so that I could find other bugs (and there are more, Sparc is coming)

Reply:
imagine the following case:

`info->code_len = 4`
`addrr = 0x1000`

reading from `m68k_read_safe_16` should now work but as info->code_len is less than the address (+2) it would now fail or am I missing something there?

Reply:
With the current code and your example, we have an out of bounds read at function `m68k_read_disassembler_16` called by `m68k_read_safe_16`, and reading `info->code[addr + 0]`

Reply:
Why? if `addr` is 0x1000 it can be a valid address to read from (0x1000 in this case isn't an actually address but an offset into the `info->code`)

Reply:
That is why we have first the computation of `addr = (address - info->baseAddress) & info->address_mask`

As you say, `addr` is rather an offset than an address.
And this offset needs to be in the bounds of the buffer.

Reply:
Yeah I guess that makes sense. As long as it works and it also doesn't trigger the issue as reporter by radare2 I'm fine with the change.